network-mux/src/Network/Mux.hs

```diff
@@ -537,3 +548,12 @@ runMiniProtocol Mux { muxMiniProtocols, muxControlCmdQueue }
   where
     ptclDir' = protocolDirEnum ptclDir
 
+    completionAction completionVar = do
+      st <- readTVar muxStatus
+      case st of
+           MuxReady -> readTMVar completionVar
+           MuxStopped  -> (readTMVar completionVar)
+             <|> (return $ Left $ toException (MuxError MuxShutdown "Mux stopped" callStack))
+           MuxFailed _ -> (readTMVar completionVar)
+             <|> (return $ Left $ toException (MuxError MuxShutdown "Mux Failed" callStack))
```

Comment:
This bit is subtle and couple probably do with some more commentary.

I think the salient point is that in the stopped or failed cases, we want to check the `readTMVar completionVar` in case it got a successful result already, but if it didn't complete yet then we want to return the mux error.

Reply:
I'll add a comment.